pkg/webhook/controlplane/controlplane.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	extensionswebhook "github.com/gardener/gardener-extensions/pkg/webhook"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+const (
+	// WebhookName is the webhook name.
+	WebhookName = "controlplane"
+)
+
+var logger = log.Log.WithName("controlplane-webhook")
+
+// AddArgs are arguments for adding a controlplane webhook to a manager.
+type AddArgs struct {
+	// Kind is the kind of this webhook
+	Kind extensionswebhook.Kind
+	// Provider is the provider of this webhook.
+	Provider string
+	// Types is a list of resource types.
+	Types []runtime.Object
+	// Validator is a validator to be used by the admission handler.
+	Validator Validator
+}
+
+// Add creates a new controlplane webhook and adds it to the given Manager.
+func Add(mgr manager.Manager, args AddArgs) (webhook.Webhook, error) {
```

Comment:
Actually, this `Add` function does not add the Webhook to the manager (or server) which is why I would either give it a different name, e.g. `CreateWebhook`, or move the logic from `webhook.AddToManager` to here.

Reply:
In fact, it does add the webhook to the manager, see the code in `webhook.NewWebhook`. It doesn't add it to the server, but I still believe there is exact symmetry with `Add` methods in controller packages, the code in `webhook.AddToManager` that calls this method is very similar for controllers and webhooks. Yet, `Add` has to return the created webhook so that it can be added to the server. The logic for creating the server and adding the webhooks to it can't be moved here, because there are multiple webhooks but just one server. We used to do it like that SEN until we added the second webhook, then we had to refactor it.

Reply:
Okay, agree. Thanks for the explanation.